### PR TITLE
feat(training): implement per-cycle distortion strength scheduling #111

### LINKED ISSUE #111: [Feature]: Per-cycle strength scheduling (paper formula)
## Problem

Paper Section 3.3 defines per-cycle distortion strength scheduling:

```
s^(c) = s_min + (s_max - s_min) * (c-1) / (C-1)
```

Where `s^(c)` is the nightmare strength at cycle `c`, `C` is total cycles. This implements curriculum adversarial training: early cycles use mild distortions, later cycles use harsh ones. The code uses a fixed `nightmare_strength: 0.8` across all cycles.

**Relevant code:** `nightmarenet/training/trainer.py` (cycle loop at line 310), `nightmarenet/data/generator.py` (creates distorted datasets with fixed strength from config), `configs/default.yaml` (`distortion.nightmare_strength: 0.8` is static).

## Proposed Solution

- In `Trainer.train()`, compute per-cycle strength using the paper's formula
- Pass the computed strength to the data generator for each cycle's dream/nightmare data
- Add config: `distortion.schedule_across_cycles: true` (enable/disable), `distortion.strength_min: 0.2`, `distortion.strength_max: 0.9`
- When disabled, behavior is unchanged (backward compatible)

## Acceptance Criteria

- [ ] Per-cycle strength computed using the paper formula
- [ ] Early cycles use lower strength, later cycles use higher
- [ ] Config `distortion.schedule_across_cycles` toggles the feature
- [ ] Default is `false` (backward compatible)
- [ ] Test: 3-cycle run produces different nightmare strengths per cycle
- [ ] Strength values logged per cycle for verification

## Scope

Medium - touches trainer loop and generator interface. ~3-4 hours.

---

> **Important:** Please read [CONTRIBUTING.md](https://github.com/Adit-Jain-srm/NightmareNet/blob/main/CONTRIBUTING.md) in full before requesting assignment. Assignment requests without a technical approach will not be considered. PRs demonstrating thorough research, benchmarks, or creative approaches beyond minimum criteria may earn bonus XP labels.

### PR DESCRIPTION
**Pull request by SanjanaTech19**
## Summary

This PR implements per-cycle distortion strength scheduling (curriculum training) for the SleepCycle training loops. Early training cycles utilize mild distortions, which progressively increase in harshness up to the maximum specified strength in later cycles.

## Motivation

Paper Section 3.3 specifies a per-cycle distortion strength schedule:
s^(c) = s_min + (s_max - s_min) * (c-1) / (C-1)
This enables curriculum training to improve model stability and general robustness, but the codebase previously used a static nightmare_strength across all training cycles.

Closes #111

## Changes

- configs/default.yaml: Added distortion.schedule_across_cycles (defaulting to false) and updated strength_min default to 0.2.
- nightmarenet/utils/config.py: Added distortion.schedule_across_cycles to configuration schema and validators, and updated default strength_min to 0.2.
- nightmarenet/training/trainer.py: Added optional arguments to Trainer.train for base datasets and generators, calculated cycle_strength at the beginning of each cycle, updated generator strengths, and dynamically regenerated/re-tokenized the dream and nightmare dataloaders per cycle.
- nightmarenet/pipeline.py: Maintained generator and base dataset instances as Pipeline properties during prepare and passed them to Trainer.train.
- tests/test_strength_scheduler.py: Created unit tests verifying scheduled strength progression, config toggle backward compatibility, and correct dataset regeneration.

## Acceptance Criteria

- [x] Per-cycle strength computed using the paper formula
- [x] Early cycles use lower strength, later cycles use higher
- [x] Config distortion.schedule_across_cycles toggles the feature
- [x] Default is false (backward compatible)
- [x] Test: 3-cycle run produces different nightmare strengths per cycle
- [x] Strength values logged per cycle for verification

## Type

- [ ] Bug fix (non-breaking change that fixes an issue)
- [x] Feature (non-breaking change that adds functionality)
- [ ] Breaking change (fix or feature that would break existing behavior)
- [ ] Refactor (no functional change)
- [ ] Documentation
- [x] Tests
- [ ] Infrastructure (CI/CD, Docker, tooling)

## Summary by CodeRabbit

* **New Features**
  * Added an option to schedule distortion strength progressively across training cycles.
  * Distortion strength can now be adjusted automatically between configured minimum and maximum values.
  * Added support for regenerating training data as distortion strength changes.

* **Configuration**
  * Added `schedule_across_cycles`, disabled by default.
  * Lowered the default minimum distortion strength from `0.3` to `0.2`.

* **Tests**
  * Added coverage for enabled and disabled distortion-strength scheduling.

### CODE REVIEW

**Review by Adit-Jain-srm — APPROVED** [review]
LGTM!
